internal/pkg/log/log.go

```diff
@@ -2,105 +2,123 @@
+	// call os.Exit.
+	Fatal(ctx context.Context, msg ...interface{})
+
+	// Sync flushes any buffered log entries.
+	Sync() error
 }
 
-// Opts allows manipulation of the Zap options.
-type Opts func(*Options)
+type logger struct {
+	zap *z.SugaredLogger
+}
 
-// addDefaults adds defaults to the Options
-func (o *Options) addDefaults() {
-	if o.OutputDest == nil {
-		o.OutputDest = os.Stderr
-	}
-	if o.Encoder == nil {
-		encCfg := zap.NewProductionEncoderConfig()
-		o.Encoder = zapcore.NewJSONEncoder(encCfg)
-	}
-	if o.Level == nil {
-		level := zap.NewAtomicLevelAt(zap.DebugLevel)
-		o.Level = &level
-	}
-	if o.StacktraceLevel == nil {
-		level := zap.NewAtomicLevelAt(zap.WarnLevel)
-		o.StacktraceLevel = &level
+// New returns an instance of Logger implemented using the Zap logging framework.
+func New(logLevel string) Logger {
+	zLevel, parseLogLevelErr := zap.ParseLogLevel(logLevel)
+
+	log := zap.New(
+		zap.Level(&zLevel),
+		// CallerSkip skips 1 number of callers, otherwise the file that gets
+		// logged will always be the wrapped file. In this case, log.go.
+		zap.AddCallerSkip(1),
+	)
+
+	if parseLogLevelErr != nil {
+		// Log an invalid log level error and set the default level to info.
+		log.Error("cannot set logger to desired log level")
 	}
+	return &logger{zap: log.Sugar()}
+}
 
-	o.ZapOptions = append(o.ZapOptions, zap.AddStacktrace(o.StacktraceLevel))
+func (l *logger) Named(name string) Logger {
+	l.zap = l.zap.Named(name)
+	return l
 }
 
-// New returns a new zap.Logger configured with the passed Options or their
-// defaults.
-func New(opts ...Opts) *zap.Logger {
-	o := &Options{}
-	for _, opt := range opts {
-		opt(o)
-	}
-	o.addDefaults()
+func (l *logger) With(args ...interface{}) Logger {
+	l.zap = l.zap.With(args...)
+	return l
+}
+
+func (l *logger) WithContext(ctx context.Context) *logger {
+	// We can add origin xDS request context here later.
+	// For now, just return the logger.
+	return l
+}
 
-	sink := zapcore.AddSync(o.OutputDest)
+func (l *logger) Sync() error { return l.zap.Sync() }
 
-	o.ZapOptions = append(o.ZapOptions, zap.AddCallerSkip(1), zap.ErrorOutput(sink))
-	log := zap.New(zapcore.NewCore(o.Encoder, sink, *o.Level))
-	log = log.WithOptions(o.ZapOptions...)
-	return log
+func (l *logger) Debug(ctx context.Context, args ...interface{}) {
+	l.WithContext(ctx).zap.Debug(args...)
 }
 
-// WriteTo configures the logger to write to the given io.Writer, instead of
-// stderr. See Options.OutputDest.
-func WriteTo(out io.Writer) Opts {
-	return func(o *Options) {
-		o.OutputDest = out
-	}
+func (l *logger) Info(ctx context.Context, args ...interface{}) {
+	l.WithContext(ctx).zap.Info(args...)
 }
 
-// Encoder configures how the logger will encode the output e.g console, JSON.
-// See Options.Encoder.
-func Encoder(encoder zapcore.Encoder) func(o *Options) {
-	return func(o *Options) {
-		o.Encoder = encoder
-	}
+func (l *logger) Warn(ctx context.Context, args ...interface{}) {
```

Comment:
I am confused about why this is not implemented in the zap.go file? Maybe I am missing something 😭

Reply:
I think there's two sides of the coin to this:
1. I can implement all these in the zap package, but the caller would be calling zap.New(), so even though they'd get the `Logger` interface back...the initial caller would still be aware of the zap package.
2. Implementing it in the log package would make it so the caller isn't aware of the zap package, but then the downside is the interface package is aware of the implementation.

I was going for 2. but I can see it working both ways, I'll go with approach 1 if you think it's cleaner.

Reply:
That is fair. Let's opt for the user experience.